fe/fe-core/src/main/java/com/starrocks/sql/InsertPlanner.java

```diff
@@ -143,11 +150,112 @@ public InsertPlanner(Map<String, Database> dbs, boolean optimisticLock) {
+        NONE_DEPEND_ON_TARGET_COLUMNS,
+        ALL_DEPEND_ON_TARGET_COLUMNS,
+        PARTIALLY_DEPEND_ON_TARGET_COLUMNS
+    }
+
+    private static GenColumnDependency getDependencyType(Column column, Set<String> targetColumns) {
+        List<SlotRef> slots = column.getGeneratedColumnRef();
+        if (slots.isEmpty()) {
+            return GenColumnDependency.NO_DEPENDENCY;
+        }
+        boolean allDependOnTargetColumns = true;
+        boolean noneDependOnTargetColumns = true;
+        for (SlotRef slot : slots) {
+            String originName = slot.getColumnName().toLowerCase();
+            if (targetColumns.contains(originName)) {
+                noneDependOnTargetColumns = false;
+            } else {
+                allDependOnTargetColumns = false;
+            }
+        }
+        if (allDependOnTargetColumns) {
+            return GenColumnDependency.ALL_DEPEND_ON_TARGET_COLUMNS;
+        }
+        if (noneDependOnTargetColumns) {
+            return GenColumnDependency.NONE_DEPEND_ON_TARGET_COLUMNS;
+        }
+        return GenColumnDependency.PARTIALLY_DEPEND_ON_TARGET_COLUMNS;
+    }
+
+    private void inferOutputSchemaForPartialUpdate(InsertStmt insertStmt) {
+        outputBaseSchema = new ArrayList<>();
+        outputFullSchema = new ArrayList<>();
+        Set<String> legalGeneratedColumnDependencies = new HashSet<>();
+        Set<String> outputColumnNames = insertStmt.getTargetColumnNames().stream()
+                .map(String::toLowerCase)
+                .collect(Collectors.toSet());
+        Set<String> baseSchemaNames = insertStmt.getTargetTable().getBaseSchema().stream()
+                .map(Column::getName)
+                .map(String::toLowerCase)
+                .collect(Collectors.toSet());
+        OlapTable targetTable = (OlapTable) insertStmt.getTargetTable();
+        for (Column column : targetTable.getFullSchema()) {
+            String columnName = column.getName().toLowerCase();
+            if (outputColumnNames.contains(columnName)) {
+                if (baseSchemaNames.contains(columnName)) {
+                    outputBaseSchema.add(column);
+                }
+                outputFullSchema.add(column);
+                legalGeneratedColumnDependencies.add(columnName);
+                continue;
+            }
+            if (column.isAutoIncrement()) {
+                if (baseSchemaNames.contains(columnName)) {
+                    outputBaseSchema.add(column);
+                }
+                outputFullSchema.add(column);
+                continue;
+            }
+            if (column.isGeneratedColumn()) {
+                // check if the generated column only depends on the columns in the output schema
+                // if so, add it to the output schema
+                // if is not related to target columns at all, skip it (TODO in future)
+                // else raise error
+                switch (getDependencyType(column, legalGeneratedColumnDependencies)) {
```

Comment:
are generated columns only at the end of schema？otherwise `legalGeneratedColumnDependencies` is incomplete.

Reply:
yes